### ISSUE TITLE
Behavior Trees by Example. AI in an Android game.

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7770]
Behavior Trees by Example. AI in an Android game.<br>
<br>
https://ift.tt/2QkW7d1<br>
<br>
<br>
<br>
<div>
<div>
<p>I'm going to cover the implementation of a very useful technique for AI programmers called Behavior Trees (BT for now on). I've been using this technique for the AI in the Android game I'm working on, and found a few quirks in the architecture that made me have to struggle somewhat. Just thought I'd share what I've learnt during the process.</p>
<p>As this is for an Android game, the code is in Java. Even so, it is easily ported to C++, and there are almost no Android specific libraries or function calls to worry about.</p>
<h2>What are Behavior Trees?</h2>
<p>The internet is crawling with places to find a good theory on what behavior trees are, and how to implement them, for example:</p>
<ul>
<li>AIGameDev : <a href="http://aigamedev.com/open/articles/bt-overview/">Overview</a>; <a href="http://aigamedev.com/open/articles/popular-behavior-tree-design/">Approaches</a>; Video tutorial <a href="http://aigamedev.com/open/articles/behavior-trees-part1/">part 1</a>, <a href="http://aigamedev.com/open/articles/behavior-trees-part2/">part 2</a>.</li>
<li>
<a href="http://en.wikipedia.org/wiki/Behavior_Trees">Wikipedia</a> (Not recommended for videogame development, but interesting)</li>
<li>Etc...</li>
</ul>
<h2>Overview</h2>
<p>This article assumes you know what a BT is. If not have a look at the mentioned articles. I&rsquo;m going to work with a system with the following architecture.</p>
<p><a href="http://lh6.ggpht.com/_vaakdclE4vY/TQfIXD96tZI/AAAAAAAABJg/vExnkX40jds/s1600-h/ClassDiagramBT%5B12%5D.png"><img border="0" src="http://lh5.ggpht.com/_vaakdclE4vY/TQfIX3ruauI/AAAAAAAABJk/r5XGZlYh5bY/ClassDiagramBT_thumb%5B8%5D.png?imgmax=800" alt="ClassDiagramBT" width="294" title="ClassDiagramBT" height="424"></a></p>
<p>It may look complex at a glance, but once we&rsquo;ve gone over each part you will see like it&rsquo;s quite understandable, elegant and flexible. At a glance you can see 3 &ldquo;groups&rdquo; of classes. The Tasks (yellow), the TaskControllers (blue) and the Decorators (green). We shall consider each of these in turn, and then how they all fit together.</p>
<h2>General</h2>
<p>The general idea for the system is that you have a set of modular tasks (MoveTo, FindClosestEnemy, FindFleeDirection, WaitTillArrivedAtDestination&hellip;) and you use them to form a BT. The base Task class provides a interface for all these tasks, the leaf tasks are the ones just mentioned, and the parent tasks are the interior nodes that decide which task to execute next.</p>
<p>The tasks have only the logic they need to actually do what is required of them, all the decision logic of whether a task has started or not, if it needs to update, if it has finished with success, etc. is grouped in the TaskController class, and added by composition. This is done for two reasons, one is for elegance when creating new tasks, the other for security when creating new decorators.</p>
<p>The decorators are tasks that &ldquo;decorate&rdquo; another class by wrapping over it and giving it additional logic. In our case we use them to give a update speed to a certain task, or to reset it once is done, or similar things. You can read about the Decorator pattern <a href="http://www.oodesign.com/decorator-pattern.html">here</a>.</p>
<p>Finally, the Blackboard class is a class owned by the parent AI that every task has a reference to. It works as a knowledge database for all the leaf tasks, and it&rsquo;s where we keep the data that is generated on one task and needs to be used by another.</p>
<h2>Tasks</h2>
<h3>Task</h3>
<p>All tasks follow the same interface:</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Base abstract class for all the tasks in</span></li>
<li><span>* the behavior tree.</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> <span>class</span> Task</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Reference to the Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>protected</span> Blackboard bb;</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the Task class</span></li>
<li><span>* @param blackboard Reference to the</span></li>
<li><span>* AI Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> Task(Blackboard blackboard)</li>
<li>{</li>
<li>
<span>this</span>.bb = blackboard;</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Override to do a pre-conditions check to</span></li>
<li><span>* see if the task can be updated.</span></li>
<li><span>* @return True if it can, false if it can't</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> boolean CheckConditions();</li>
<li><span>/**</span></li>
<li><span>* Override to add startup logic to the task</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> <span>void</span> Start();</li>
<li><span>/**</span></li>
<li><span>* Override to add ending logic to the task</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> <span>void</span> End();</li>
<li><span>/**</span></li>
<li><span>* Override to specify the logic the task</span></li>
<li><span>* must update each cycle</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> <span>void</span> DoAction();</li>
<li><span>/**</span></li>
<li><span>* Override to specify the controller the</span></li>
<li><span>* task has</span></li>
<li><span>* @return The specific task controller.</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> TaskController GetControl();</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<p>As in any BT node, a CheckConditions and a DoAction functions, to check if the node can be updated, and to actually update the node, respectively. The Start and End functions are called just before starting to update the node, and just after finishing the logic of the node.</p>
<p>To be a true interface, the Blackboard member should not be in the class, but for simplicity and ease of use, it&rsquo;s the perfect place for it. So fuck interfaces then.</p>
<h3>Leaf Task</h3>
<h4>Base LeafTask class</h4>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Leaf task (or node) in the behavior tree.</span></li>
<li><span>*</span></li>
<li><span>* Specifies a TaskControler, by composition,</span></li>
<li><span>* to take care of all the control logic,</span></li>
<li><span>* without burdening the Task class with</span></li>
<li><span>* complications.</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> <span>class</span> LeafTask extends Task</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Task controler to keep track of the</span></li>
<li><span>* Task state.</span></li>
<li><span>*/</span></li>
<li>
<span>protected</span> TaskController control;</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* LeafTask class</span></li>
<li><span>* @param blackboard Reference to the</span></li>
<li><span>* AI Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> LeafTask(Blackboard blackboard)</li>
<li>{</li>
<li>super(blackboard);</li>
<li>CreateController();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Creates the controller for the class</span></li>
<li><span>*/</span></li>
<li>
<span>private</span> <span>void</span> CreateController()</li>
<li>{</li>
<li>
<span>this</span>.control = <span>new</span> TaskController(<span>this</span>);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Gets the controller reference.</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> TaskController GetControl()</li>
<li>{</li>
<li>
<span>return</span> <span>this</span>.control;</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<p>The leaf task has a TaskController for the support logic of the class, and implements the GetControl method. It has no more extra logic or methods.</p>
<p>The child classes of the LeafTask class are the ones with the real logic of the system, and each time we want a new behavior we are going to have to create a few. They are building blocks, modular, and they can be combined in different ways to generate different effects. Take for example simplistic Chase and a Flee strategies, they could work like this:</p>
<p>Chase</p>
<ol>
<li>GetClosestEnemy</li>
<li>SetClosestEnemyPositionAsDestination</li>
<li><em>MoveToDestination</em></li>
<li><em>WaitUntilPositionNearDestination</em></li>
</ol>
<p>Flee</p>
<ol>
<li>CalculateFleeDestination</li>
<li><em>MoveToDestination</em></li>
<li><em>WaitUntilPositionNearDestination</em></li>
</ol>
<p>The italics mark the reused leaf nodes in two different sequences of behaviors. That shows how modular leaf nodes can be recycled and used to conform many different behaviors.</p>
<p>Leaf tasks often calculate data needed by other leaf tasks. Instead of making a complex linking of the whole tree of tasks, we just use a common &ldquo;data dump&rdquo; for all of them, a Blackboard object shared by all the tasks, that simply has member data used by the different tasks.</p>
<p>Some examples of Leaf Nodes follow.</p>
<h4>LeafTask Examples</h4>
<h5>GetClosestEnemyTask</h5>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Finds the closest enemy and stores</span></li>
<li><span>* it in the Blackboard.</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> GetClosestEnemyTask extends LeafTask</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* GetClosestEnemyTask task</span></li>
<li><span>* @param blackboard Reference to the</span></li>
<li><span>* AI Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> GetClosestEnemyTask(Blackboard bb)</li>
<li>{</li>
<li>super(bb);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Checks for preconditions</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> boolean CheckConditions()</li>
<li>{</li>
<li>
<span>return</span> Blackboard.players.size() &gt; 1 &amp;&amp;</li>
<li>bb.player.GetCursor().GetPosition() != null;</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Finds the closest enemy and stores</span></li>
<li><span>* it in the Blackboard</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> DoAction()</li>
<li>{</li>
<li>Enemy closestEnemy = null;</li>
<li><span>// Finds the closest enemy</span></li>
<li><span>// Omited for clarity</span></li>
<li><span>// Store it in the blackboard</span></li>
<li>bb.closestEnemy = closestEnemy;</li>
<li>GetControll().FinishWithSuccess();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Ends the task</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> End()</li>
<li>{</li>
<li>LogTask(<span>"Ending"</span>);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Starts the task</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> Start()</li>
<li>{</li>
<li>LogTask(<span>"Starting"</span>);</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h5>MoveTo</h5>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* This task requests the player to</span></li>
<li><span>* move to a given destination.</span></li>
<li><span>* It takes the destination from</span></li>
<li><span>* the Blackboard.</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> MoveToDestinationTask extends LeafTask</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* MoveToDestinationTask class</span></li>
<li><span>* @param blackboard Reference of the</span></li>
<li><span>* AI Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> MoveToDestinationTask(Blackboard bb)</li>
<li>{</li>
<li>super(bb);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Sanity check of needed data for the</span></li>
<li><span>* operations</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> boolean CheckConditions()</li>
<li>{</li>
<li>LogTask(<span>"Checking conditions"</span>);</li>
<li>
<span>return</span> bb.moveDirection != null;</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Requests the cursor to move</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> DoAction()</li>
<li>{</li>
<li>LogTask(<span>"Doing action"</span>);</li>
<li>bb.player.Move(bb.moveDirection);</li>
<li>control.FinishWithSuccess();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Ends the task</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> End()</li>
<li>{</li>
<li>LogTask(<span>"Ending"</span>);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Starts the task</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> Start()</li>
<li>{</li>
<li>LogTask(<span>"Starting"</span>);</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h3>Parent Task</h3>
<p>Parent task require a bit more explanation than leaf tasks, but don&rsquo;t worry, we&rsquo;ll work through it step by step. The base class is similar to the LeafTask class:</p>
<p>The ParentTask class has a ParentTaskController, identical to the TaskController for the LeafNode (handles state logic, if it&rsquo;s ready, finished, ect&hellip;) but adds responsibility for the child tasks (a subtasks vector) and the current selected task (curTask).</p>
<p>It also implements the Start and End functions, very similar to the LeafTask. But the big change comes in the DoAction function. Here it updates the child of the parent task acordingly. All Parent tasks have the same process, that can be explained conceptually more or less like this:</p>
<ol>
<li>Safety checks to see if we have all the data, and it&rsquo;s not null</li>
<li>If the current child task is not started, start it</li>
<li>Else, if the current child task is finished, finalize it propperly and check if it finished with success or failure. <strong>Here</strong> is where the different ParentTask differ, on what happens if the child fails or succeeds. (ChildSuceeded and ChildFailed virtual functions)</li>
<li>Else, the child must be updated, so we call it&rsquo;s DoAction.</li>
</ol>
<p>The abstract ChildSuceeded and ChildFailed functions are for the classes derived from ParentTask to specify how they behave when a child ends their execution. We have two distinct cases for this, Sequence and Selector classes.</p>
<h4>Sequence</h4>
<p>In a sequence of tasks, if a child ends it&rsquo;s execution with a <strong>failure</strong>, we <strong>bail</strong> and the Sequence ParentTask ends with failure. This makes sense, as sequences are used for example for the Chase explained earlier:</p>
<ol>
<ol>
<li>GetClosestEnemy</li>
<li>SetClosestEnemyPositionAsDestination</li>
<li><em>MoveToDestination</em></li>
<li><em>WaitUntilPositionNearDestination</em></li>
</ol>
</ol>
<p>If task 2 fails, we don&rsquo;t want the ParentTask to continue on to MoveToDestination, as the Destination Vec2 is possibly corrupted, and will result in undefined behavior.</p>
<p>On the other hand if a child finishes with success, we continue onto the next task.</p>
<p>Some code to illustrate this:</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* This ParentTask executes each of it's children</span></li>
<li><span>* in turn until he has finished all of them.</span></li>
<li><span>*</span></li>
<li><span>* It always starts by the first child,</span></li>
<li><span>* updating each one.</span></li>
<li><span>* If any child finishes with failure, the</span></li>
<li><span>* Sequence fails, and we finish with failure.</span></li>
<li><span>* When a child finishes with success, we</span></li>
<li><span>* select the next child as the update victim.</span></li>
<li><span>* If we have finished updating the last child,</span></li>
<li><span>* the Sequence returns with success.</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> Sequence extends ParentTask</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* Sequence class</span></li>
<li><span>* @param blackboard Reference to</span></li>
<li><span>* the AI Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> Sequence(Blackboard bb)</li>
<li>{</li>
<li>super(bb);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* A child finished with failure.</span></li>
<li><span>* We failed to update the whole sequence.</span></li>
<li><span>* Bail with failure.</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> ChildFailed()</li>
<li>{</li>
<li>control.FinishWithFailure();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* A child has finished with success</span></li>
<li><span>* Select the next one to update. If</span></li>
<li><span>* it's the last, we have finished with</span></li>
<li><span>* success.</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> ChildSucceeded()</li>
<li>{</li>
<li>
<span>int</span> curPos =</li>
<li>control.subtasks.</li>
<li>indexOf(control.curTask);</li>
<li>
<span>if</span>( curPos ==</li>
<li>(control.subtasks.size() - 1))</li>
<li>{</li>
<li>control.FinishWithSuccess();</li>
<li>}</li>
<li><span>else</span></li>
<li>{</li>
<li>control.curTask =</li>
<li>control.subtasks.</li>
<li>elementAt(curPos+1);</li>
<li>
<span>if</span>(!control.curTask.CheckConditions())</li>
<li>{</li>
<li>control.FinishWithFailure();</li>
<li>}</li>
<li>}</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h4>Selector</h4>
<p>The selector chooses one it&rsquo;s children to update, and if that fails it chooses another until there are no more left. This means that if the current chosen child ends with failure, we choose another. If we can&rsquo;t choose another, we end with failure. On the other hand, if our child returns with success, we can happily finish with success as well, as we only need one of the children to succeed.</p>
<p>In retrospect, the Sequencer is like a AND gate, and the Selector as a OR gate, only applied to tasks instead of circuits&hellip;</p>
<p>Some code then.</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* This parent task selects one of it's</span></li>
<li><span>* children to update.</span></li>
<li><span>*</span></li>
<li><span>* To select a child, it starts from the</span></li>
<li><span>* beginning of it's children vector</span></li>
<li><span>* and goes one by one until it finds one</span></li>
<li><span>* that passes the CheckCondition test.</span></li>
<li><span>* It then updates that child until its</span></li>
<li><span>* finished.</span></li>
<li><span>* If the child finishes with failure,</span></li>
<li><span>* it continues down the list looking another</span></li>
<li><span>* candidate to update, and if it doesn't</span></li>
<li><span>* find it, it finishes with failure.</span></li>
<li><span>* If the child finishes with success, the</span></li>
<li><span>* Selector considers it's task done and</span></li>
<li><span>* bails with success.</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> Selector extends ParentTask</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of</span></li>
<li><span>* the Selector class</span></li>
<li><span>* @param blackboard Reference to</span></li>
<li><span>* the AI Blackboard data</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> Selector(Blackboard bb)</li>
<li>{</li>
<li>super(bb);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Chooses the new task to update.</span></li>
<li><span>* @return The new task, or null</span></li>
<li><span>* if none was found</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> Task ChooseNewTask()</li>
<li>{</li>
<li>Task task = null;</li>
<li>boolean found = <span>false</span>;</li>
<li>
<span>int</span> curPos =</li>
<li>control.subtasks.</li>
<li>indexOf(control.curTask);</li>
<li>
<span>while</span>(!found)</li>
<li>{</li>
<li>
<span>if</span>(curPos ==</li>
<li>(control.subtasks.size() - 1))</li>
<li>{</li>
<li>found = <span>true</span>;</li>
<li>task = null;</li>
<li>
<span>break</span>;</li>
<li>}</li>
<li>curPos++;</li>
<li>task = control.subtasks.</li>
<li>elementAt(curPos);</li>
<li>
<span>if</span>(task.CheckConditions())</li>
<li>{</li>
<li>found = <span>true</span>;</li>
<li>}</li>
<li>}</li>
<li>
<span>return</span> task;</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* In case of child finishing with</span></li>
<li><span>* failure we find a new one to update,</span></li>
<li><span>* or fail if none is to be found</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> ChildFailed()</li>
<li>{</li>
<li>control.curTask = ChooseNewTask();</li>
<li>
<span>if</span>(control.curTask == null)</li>
<li>{</li>
<li>control.FinishWithFailure();</li>
<li>}</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* In case of child finishing with</span></li>
<li><span>* sucess, our job here is done, finish</span></li>
<li><span>* with sucess</span></li>
<li><span>* as well</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> ChildSucceeded()</li>
<li>{</li>
<li>control.FinishWithSuccess();</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h2>Task Controller</h2>
<h3>Introduction</h3>
<p>The task controller, as mentioned earlier, tracks the state of the Task it is added to. It keeps a reference to the task and acts as a wrapper for said class when dealing with all the &ldquo;Is it finished? Is it ready to update?&rdquo; kind of questions.</p>
<p>This code is separated from the actual task for two reasons.</p>
<h4>Reason One: Elegance of Design.</h4>
<p>Once we separate the state control from the task class, we can create new subtasks paying no mind to the whole mechanism going on in the background to keep our task in harmony with it&rsquo;s parent and children tasks. In the Task classes we just focus on specifying the specific logic for the task at hand.</p>
<h4>Reason Two: Safety in the Decorators</h4>
<p>If we take the utility/state logic out of the Task class, we can make all the methods abstract, and when we create the Decorator classes, the compiler checks to see if we have overridden all the abstract functions (if we forget to wrap any of the Task functions in the Decorator, we could introduce subtle but dangerous bugs). If we have the utility functions in the Task class we may or may not remember to make a wrapper for them, and if we have to change the interface of the Task class at some point, we could be in for a hell of pain.</p>
<h3>Code</h3>
<p>All said, some code on how the task controller works:</p>
<h4>TaskController</h4>
<p>Has the <strong>done</strong> and <strong>success</strong> flags, and all the utilities related with them.</p>
<h4>ParentTaskController</h4>
<p>We add the subtasks vector, and curTask task to the previous responsibilities of the taskController.</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* This class extends the TaskController</span></li>
<li><span>* class to add support for</span></li>
<li><span>* child tasks and their logic.</span></li>
<li><span>*</span></li>
<li><span>* Used together with ParentTask.</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> ParentTaskController extends TaskController</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Vector of child Task</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> Vector&lt;Task&gt; subtasks;</li>
<li><span>/**</span></li>
<li><span>* Current updating task</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> Task curTask;</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* ParentTaskController class</span></li>
<li><span>* @param task</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> ParentTaskController(Task task)</li>
<li>{</li>
<li>super(task);</li>
<li>
<span>this</span>.subtasks = <span>new</span> Vector&lt;Task&gt;();</li>
<li>
<span>this</span>.curTask = null;</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Adds a new subtask to the end</span></li>
<li><span>* of the subtask list.</span></li>
<li><span>* @param task Task to add</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>void</span> Add(Task task)</li>
<li>{</li>
<li>subtasks.add(task);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Resets the task as if it had</span></li>
<li><span>* just started.</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>void</span> Reset()</li>
<li>{</li>
<li>super.Reset();</li>
<li>
<span>this</span>.curTask =</li>
<li>subtasks.firstElement();</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h2>Decorator</h2>
<p>The Decorators are used in the BT to add special functionality to any given task. They act as wrappers of the class, calling the class methods and adding the extra functionality where they deem it necessary.</p>
<h3>Abstract Decorator Class</h3>
<p>A Decorator, as explained in the <a href="http://www.oodesign.com/decorator-pattern.html">Decorator Pattern</a> has a reference to the class it &ldquo;decorates&rdquo; (in our case <em>private Task task</em>), and also inherits from said class.</p>
<p>Normally, the Decorator adds extra logic in the DoAction method of the task, which is why the base Decorator class presented only overrides the rest of the methods of Task, for simplicity. None the less, if any Decorator subclass needs to override any other method, it can do so with no problem.</p>
<p>Here is the code</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Base class for the specific decorators.</span></li>
<li><span>* Decorates all the task methods except</span></li>
<li><span>* for the DoAction, for commodity.</span></li>
<li><span>*</span></li>
<li><span>* (Tough any method can be decorated in</span></li>
<li><span>* the base classes with no problem,</span></li>
<li><span>* they are decorated by default so the</span></li>
<li><span>* programmer does not forget)</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>abstract</span> <span>class</span> TaskDecorator extends Task</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Reference to the task to decorate</span></li>
<li><span>*/</span></li>
<li>Task task;</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* Decorator class</span></li>
<li><span>* @param blackboard Reference to</span></li>
<li><span>* the AI Blackboard data</span></li>
<li><span>* @param task Task to decorate</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> TaskDecorator(Blackboard bb, Task task)</li>
<li>{</li>
<li>super(bb);</li>
<li>InitTask(task);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Initializes the task reference</span></li>
<li><span>* @param task Task to decorate</span></li>
<li><span>*/</span></li>
<li>
<span>private</span> <span>void</span> InitTask(Task task)</li>
<li>{</li>
<li>
<span>this</span>.task = task;</li>
<li>
<span>this</span>.task.GetControl().SetTask(<span>this</span>);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Decorate the CheckConditions</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> boolean CheckConditions()</li>
<li>{</li>
<li>
<span>return</span> <span>this</span>.task.CheckConditions();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Decorate the end</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> End()</li>
<li>{</li>
<li>
<span>this</span>.task.End();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Decorate the request for the</span></li>
<li><span>* Controll reference</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> TaskController GetControl()</li>
<li>{</li>
<li>
<span>return</span> <span>this</span>.task.GetControl();</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Decorate the start</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> Start()</li>
<li>{</li>
<li>
<span>this</span>.task.Start();</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h3>Examples</h3>
<p>Here are some specific examples of Decorators.</p>
<h4>ResetDecorator</h4>
<p>Simply checks to see if the task it decorates has finished. If it has, it resets the task to &ldquo;ready to execute again&rdquo;.</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Decorator that resets to "Started" the task</span></li>
<li><span>* it is applied to, each time said task finishes.</span></li>
<li><span>*</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> ResetDecorator extends TaskDecorator</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* ResetDecorator class</span></li>
<li><span>* @param blackboard Reference to</span></li>
<li><span>* the AI Blackboard data</span></li>
<li><span>* @param task Task to decorate</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> ResetDecorator(Blackboard bb,</li>
<li>Task task)</li>
<li>{</li>
<li>super(bb, task);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Does the decorated task's action,</span></li>
<li><span>* and if it's done, resets it.</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> DoAction()</li>
<li>{</li>
<li>
<span>this</span>.task.DoAction();</li>
<li>
<span>if</span>(<span>this</span>.task.GetControl().Finished())</li>
<li>{</li>
<li>
<span>this</span>.task.GetControl().Reset();</li>
<li>}</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h4>RegulatorDecorator</h4>
<p>Updates the task it decorates at a specific speed. This case of Decorator also overrides the Start method.</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Decorator that adds a update speed</span></li>
<li><span>* limit to the task it is applied to</span></li>
<li><span>* @author Ying</span></li>
<li><span>*</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> <span>class</span> RegulatorDecorator extends TaskDecorator</li>
<li>{</li>
<li><span>/**</span></li>
<li><span>* Regulator to keep track of time</span></li>
<li><span>*/</span></li>
<li>
<span>private</span> Regulator regulator;</li>
<li><span>/**</span></li>
<li><span>* Update time in seconds per frame</span></li>
<li><span>*/</span></li>
<li>
<span>private</span> <span>float</span> updateTime;</li>
<li><span>/**</span></li>
<li><span>* Creates a new instance of the</span></li>
<li><span>* RegulatorDecorator class</span></li>
<li><span>* @param blackboard Reference to the</span></li>
<li><span>* AI Blackboard data</span></li>
<li><span>* @param task Task to decorate</span></li>
<li><span>* @param updateTime Time between each</span></li>
<li><span>* frame update</span></li>
<li><span>*/</span></li>
<li>
<span>public</span> RegulatorDecorator(Blackboard bb,</li>
<li>Task task, <span>float</span> updateTime)</li>
<li>{</li>
<li>super(bb, task);</li>
<li>
<span>this</span>.updateTime = updateTime;</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Starts the task and the regulator</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> Start()</li>
<li>{</li>
<li>task.Start();</li>
<li>
<span>this</span>.regulator =</li>
<li>
<span>new</span> Regulator(1.0f/updateTime);</li>
<li>}</li>
<li><span>/**</span></li>
<li><span>* Updates the decorated task only if the</span></li>
<li><span>* required time since the last update has</span></li>
<li><span>* elapsed.</span></li>
<li><span>*/</span></li>
<li>@Override</li>
<li>
<span>public</span> <span>void</span> DoAction()</li>
<li>{</li>
<li>
<span>if</span>(<span>this</span>.regulator.IsReady())</li>
<li>{</li>
<li>task.DoAction();</li>
<li>}</li>
<li>}</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h2>Usage Example</h2>
<p>With all the code mentioned earlier, it&rsquo;s a little easy to get lost, so here is a reminder of how this is supposed to work. We are going to create a simple BT with just two basic behaviors, Chase and Flee. (This is taken straight from the game code).</p>
<div>
<div>
<p>Code Snippet</p>
<div>
<ol>
<li><span>/**</span></li>
<li><span>* Creates the behavior tree and populates</span></li>
<li><span>* the node hierarchy</span></li>
<li><span>*/</span></li>
<li>
<span>private</span> <span>void</span> CreateBehaviourTree()</li>
<li>{</li>
<li><span>// Create a root node for the tree, that</span></li>
<li><span>// resets itself and updates at 10 fps</span></li>
<li>
<span>this</span>.planner = <span>new</span> Selector(</li>
<li>blackboard,</li>
<li>
<span>"Planner"</span>);</li>
<li>
<span>this</span>.planner = <span>new</span> ResetDecorator(</li>
<li>blackboard,</li>
<li>
<span>this</span>.planner, <span>"Planner"</span>);</li>
<li>
<span>this</span>.planner = <span>new</span> RegulatorDecorator(</li>
<li>blackboard,</li>
<li>
<span>this</span>.planner, <span>"Planner"</span>, 0.1f);</li>
<li><span>// Create chase sequence</span></li>
<li>Task chase = <span>new</span> Sequence(</li>
<li>blackboard,</li>
<li>
<span>"Chase sequence"</span>);</li>
<li>((ParentTaskController)chase.GetControl()).</li>
<li>Add(<span>new</span> GetClosestEnemyCursorTask(</li>
<li>blackboard,</li>
<li>
<span>"GetClosestEnemyCursor"</span>));</li>
<li>((ParentTaskController)chase.GetControl()).</li>
<li>Add(<span>new</span> SetEnemyCursorAsDestinationTask(</li>
<li>blackboard,</li>
<li>
<span>"SetEnemyCursorAsDestination"</span>));</li>
<li>((ParentTaskController)chase.GetControl()).</li>
<li>Add(<span>new</span> MoveToDestinationTask(</li>
<li>blackboard,</li>
<li>
<span>"MoveToDestination"</span>));</li>
<li>((ParentTaskController)chase.GetControl()).</li>
<li>Add(<span>new</span> WaitTillNearDestinationTask(</li>
<li>blackboard,</li>
<li>
<span>"WaitTillNearDestination"</span>));</li>
<li><span>// Create the flee sequence</span></li>
<li><span>// It's a normal selector but with extra logic</span></li>
<li><span>// to see if we want to flee or not</span></li>
<li>Task flee = <span>new</span> Sequence(blackboard,</li>
<li>
<span>"Flee sequence"</span>);</li>
<li>flee = <span>new</span> FleeDecorator(blackboard, flee,</li>
<li>
<span>"Flee sequence"</span>);</li>
<li>((ParentTaskController)flee.GetControl()).</li>
<li>Add(<span>new</span> CalculateFleeDestinationTask(</li>
<li>blackboard,</li>
<li>
<span>"CalculateFleeDestination"</span>));</li>
<li>((ParentTaskController)flee.GetControl()).</li>
<li>Add(<span>new</span> MoveToDestinationTask(</li>
<li>blackboard,</li>
<li>
<span>"MoveToDestination"</span>));</li>
<li>((ParentTaskController)flee.GetControl()).</li>
<li>Add(<span>new</span> WaitTillNearDestinationTask(</li>
<li>blackboard,</li>
<li>
<span>"WaitTillNearDestination"</span>));</li>
<li>((ParentTaskController)<span>this</span>.planner.</li>
<li>GetControl()).Add(flee);</li>
<li>((ParentTaskController)<span>this</span>.planner.</li>
<li>GetControl()).Add(chase);</li>
<li>}</li>
</ol>
</div>
</div>
</div>
<h2>Diagram</h2>
<p>The previous code generates the Behavior Tree shown in this diagram.</p>
<p><a href="http://lh3.ggpht.com/_vaakdclE4vY/TQfIYaVmvpI/AAAAAAAABJo/0SErS9srhQs/s1600-h/FlowDiagramBT%5B4%5D.png"><img border="0" src="http://lh4.ggpht.com/_vaakdclE4vY/TQfIZVEXueI/AAAAAAAABJs/41gOyssQkSs/FlowDiagramBT_thumb%5B2%5D.png?imgmax=800" alt="FlowDiagramBT" width="326" title="FlowDiagramBT" height="549"></a></p>
<p>Behavior trees are a incredibly interesting tool for game AI. They can be applied to many different types of AI and their modularity makes them a blessing when it comes to extending or modifying an existing AI.</p>
<p>I just hope that with this huge post I have provided the necessary tools for someone who is looking for a nice AI technique for his AI to start working and hopefully avoid the pitfalls I had to survive whilst doing this system.</p>
<p>Happy coding.</p>
<p>PS: The full code for my game Behavior Tree can be found <a href="http://code.google.com/p/daggame/source/browse/#svn/trunk/Code/src/com/game/AI">here</a>.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via magicscrollsofcode.blogspot.com https://ift.tt/3tgFQEr<br>
<br>
April 14, 2021 at 11:21AM
